### PR TITLE
reminders with iso 8601 skeleton implementation

### PR DESCRIPTION
**Pull request by jigargandhi**
# Description

Added support for number of repetitions for a reminder 

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #702 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by jigargandhi** [comment]
@dapr/maintainers-dotnet-sdk Support for reminders is added through actor classes implementing `IRemindable` interface which has method `Task ReceiveReminderAsync(string, byte[] , TimeSpan , TimeSpan )`, Now with repetitions and ISO 8601 intervals reminders may run only for a limited number of times. 
Do you guys think, we should include the attempts remaining in the interface method?
If yes, adding it to the method will be a breaking change. So we can think of a struct/class holding this properties.

File: src/Dapr.Actors/Runtime/ConverterUtils.cs [hunk 1]
```diff
@@ -60,5 +63,69 @@ public static string ConvertTimeSpanValueInDaprFormat(TimeSpan? value)
 
             return stringValue;
         }
+
+        public static string ConvertTimeSpanValueInISO8601Format(TimeSpan value, int repetitions)
```
Comment (halspang) [inline src/Dapr.Actors/Runtime/ConverterUtils.cs:67]: A lot of this method is saying it isn't covered, can we get tests for it?

File: src/Dapr.Actors/Runtime/ConverterUtils.cs [hunk 2]
```diff
@@ -60,5 +63,69 @@ public static string ConvertTimeSpanValueInDaprFormat(TimeSpan? value)
...
+        {
+            if (repetitions == -1)
+            {
+                return ConvertTimeSpanValueInDaprFormat(value);
+            }
```
Comment (halspang) [inline src/Dapr.Actors/Runtime/ConverterUtils.cs:72]: Why the separate path? Why not just use the repetition as normal?

File: src/Dapr.Actors/Runtime/ConverterUtils.cs (hunk 1 above)
Comment (halspang) [inline src/Dapr.Actors/Runtime/ConverterUtils.cs:67]: Is it possible for someone to pass in a TimeSpan with a repetition and one as a method input? If so, which one wins? Or do we error?

File: src/Dapr.Actors/Runtime/ActorReminder.cs [hunk 3]
```diff
@@ -58,6 +58,33 @@ public class ActorReminder : ActorReminderToken, IActorReminder
...
+            byte[] state,
+            TimeSpan dueTime,
+            TimeSpan period,
+            int repetitions)
+            : this(actorType, actorId, name, state, dueTime, period)
+        {
+            if (repetitions < 0)
+            {
+                throw new ArgumentException("TODO: fill in message");
+            }
```
[the remark this answers is not in the document]
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/ActorReminder.cs:84]: We only support positive values for invocations.

File: src/Dapr.Actors/Runtime/Actor.cs [hunk 4]
```diff
@@ -213,7 +213,38 @@ protected virtual Task OnPostActorMethodAsync(ActorMethodContext actorMethodCont
...
+            TimeSpan period,
+            int repetitions)
+        {
+            var reminder = new ActorReminder(this.actorTypeName, this.Id, reminderName, state, dueTime, period, repetitions);
+            await this.Host.TimerManager.RegisterReminderAsync(reminder);
+            return reminder;
+        }
+
+        /// <summary>
+        /// Unregisters a reminder previously registered using <see cref="o:Dapr.Actors.Runtime.Actor.RegisterReminderAsync" />.
```
[the remark this answers is not in the document]
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/Actor.cs:247]: I copied it from existing method, I guess, this is no longer supported.

File: src/Dapr.Actors/Runtime/ConverterUtils.cs (hunk 2 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/ConverterUtils.cs:72]: There may be reminder periods still created in golang's time.Duration format. When we added the repetition to runtime, we wanted to support both time.Duration format and ISO 8601 format. hence the separate paths.

File: examples/Actor/DemoActor/DemoActor.cs [hunk 5]
```diff
@@ -71,8 +72,9 @@ public async Task ReceiveReminderAsync(string reminderName, byte[] state, TimeSp
         {
             // This method is invoked when an actor reminder is fired.
             var actorState = await this.StateManager.GetStateAsync<MyData>(StateName);
-            actorState.PropertyB = $"Reminder triggered at '{DateTime.Now:yyyy-MM-ddTHH:mm:ss}'";
+            actorState.PropertyB = $"Reminder triggered at '{DateTime.Now:yyyy-MM-ddTHH:mm:ss}' with";
             await this.StateManager.SetStateAsync<MyData>(StateName, actorState);
```
Comment (halspang) [inline examples/Actor/DemoActor/DemoActor.cs:76]: 'with'?

File: src/Dapr.Actors/Runtime/ActorReminder.cs [hunk 6]
```diff
@@ -58,6 +58,33 @@ public class ActorReminder : ActorReminderToken, IActorReminder
...
+        /// <summary>
+        /// Initializes a new instance of <see cref="ActorReminder" />.
+        /// </summary>
+        /// <param name="actorType">The actor type.</param>
+        /// <param name="actorId">The actor id.</param>
+        /// <param name="name">The reminder name.</param>
+        /// <param name="state">The state associated with the reminder.</param>
+        /// <param name="dueTime">The reminder due time.</param>
+        /// <param name="period">The reminder period.</param>
+        /// <param name="repetitions">The number of time reminder should be invoked.</param>
```
Comment (halspang) [inline src/Dapr.Actors/Runtime/ActorReminder.cs:70]: time -> times

File: src/Dapr.Actors/Runtime/ActorReminder.cs [hunk 7]
```diff
@@ -72,5 +99,10 @@ public class ActorReminder : ActorReminderToken, IActorReminder
         /// Gets the reminder period.
         /// </summary>
         public TimeSpan Period { get; }
+
+        /// <summary>
+        /// Gets the number of invocations of the reminder left.
+        /// </summary>
+        public int RepetitionsLeft { get; }
```
Comment (halspang) [inline src/Dapr.Actors/Runtime/ActorReminder.cs:106]: Do we need a default for people who call the original constructor?

File: src/Dapr.Actors/Runtime/ActorReminder.cs [hunk 8]
```diff
@@ -58,6 +58,33 @@ public class ActorReminder : ActorReminderToken, IActorReminder
...
+            byte[] state,
+            TimeSpan dueTime,
+            TimeSpan period,
+            int repetitions)
+            : this(actorType, actorId, name, state, dueTime, period)
+        {
+            if (repetitions < 0)
+            {
+                throw new ArgumentException(SR.RepetitionArgumentOutOfRange);
+            }
```
Comment (halspang) [inline src/Dapr.Actors/Runtime/ActorReminder.cs:84]: Isn't -1 a valid case?

File: test/Dapr.Actors.Test/Runtime/DefaultActorTimerManagerTests.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+namespace Dapr.Actors.Runtime
+{
+    public sealed class DefaultActorTimerManagerTests
+    {
+        /// <summary>
+        /// When register reminder is called, interactor is called with correct data.
+        /// </summary>
+        /// <returns></returns>
+        [Fact]
+        public async Task RegisterReminderAsync_CallsInteractor_WithCorrectData()
```
Comment (halspang) [inline test/Dapr.Actors.Test/Runtime/DefaultActorTimerManagerTests.cs:23]: Should we have a reminder with repetitions here as well?

File: src/Dapr.Actors/Runtime/ActorReminder.cs (hunk 7 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/ActorReminder.cs:106]: The default will be a zero which is fine.

File: src/Dapr.Actors/Runtime/ActorReminder.cs (hunk 8 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/ActorReminder.cs:84]: No, dapr api will not allow -1

File: src/Dapr.Actors/Runtime/ConverterUtils.cs (hunk 1 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/ConverterUtils.cs:67]: I am sorry, did not get your comment here

File: src/Dapr.Actors/Dapr.Actors.csproj [hunk 10]
```diff
@@ -1,26 +1,50 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
-  <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
-  </PropertyGroup>
+    <PropertyGroup>
```
Comment (rynowak) [inline src/Dapr.Actors/Dapr.Actors.csproj:3]: @jigargandhi - what are the changes in this file accomplishing? I'm not sure how this is related to adding the new API.

File: src/Dapr.Actors/Resources/SR.resx [hunk 11]
```diff
@@ -114,113 +119,154 @@
...
     <value>Method Id '{0}' for interface Id '{1}' not found in service implementation. If a new method is added to interface and client &amp; service are being upgraded at the same time, its possible that client got upgraded before the service. If a method is removed from the interface and client &amp; service are being upgraded at the same time, its possible that service got upgraded before the client. Addition or removal of methods to an interface should be performed as a phased upgrade.</value>
+    <comment/>
   </data>
   <data name="ErrorNamedActorStateNotFound" xml:space="preserve">
     <value>Actor State with name {0} was not found.</value>
+    <comment/>
   </data>
   <data name="ErrorNoServiceInterfaceFound" xml:space="preserve">
     <value>The service type '{0}' does not implement any service interfaces or one of the interfaces implemented is not a service interface. All interfaces(including its parent interface) implemented by service type must be service interface. A service interface is the one that ultimately derives from '{1}' type.</value>
+    <comment/>
```
Comment (rynowak) [inline src/Dapr.Actors/Resources/SR.resx:178]: Can you explain or revert the changes to this file?  I'm not sure what's happening here.

File: src/Dapr.Actors/Runtime/ActorReminder.cs (hunk 7 above)
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorReminder.cs:106]: Can this property be renamed to match the parameter? `Repetitions`. I also would prefer that this reflect the behavior (1 instead of 1). I don't get why it has to be 0.

File: src/Dapr.Actors/Resources/SR.resx (hunk 11 above)
Comment (rynowak) [reply src/Dapr.Actors/Resources/SR.resx:178]: Were these changes made by visual studio?

File: src/Dapr.Actors/Runtime/ConverterUtils.cs [hunk 12]
```diff
@@ -60,5 +63,69 @@ public static string ConvertTimeSpanValueInDaprFormat(TimeSpan? value)
...
+
+            if(value.Seconds > 0)
+            {
+                builder.AppendFormat("{0}S", value.Seconds);
+            }
+            return builder.ToString();
+        }
+        
+        public static (TimeSpan, int) ConvertTimeSpanValueFromISO8601Format(string valueString)
```
Comment (rynowak) [inline src/Dapr.Actors/Runtime/ConverterUtils.cs:101]: If this is used needed for test code then this **must** live in the tests. Regexes in production code need to be scrutinized **heavily** for security reasons.

File: src/Dapr.Actors/Runtime/IActorReminder.cs [hunk 13]
```diff
@@ -41,5 +41,11 @@ public interface IActorReminder
         /// </summary>
         /// <value>The user state passed to the reminder invocation.</value>
         byte[] State { get; }
+
+        /// <summary>
+        /// Gets the number of repetitions of the reminder left to be invoked.
+        /// </summary>
+        /// <value>The number of repetitions of the reminder left to be invoked.</value>
+        int RepetitionsLeft { get; }
```
Comment (rynowak) [inline src/Dapr.Actors/Runtime/IActorReminder.cs:49]: This is a binary breaking change, we cannot accept it, so we need to find another solution. Is there anywhere that this property is read?

File: src/Dapr.Actors/Runtime/ConverterUtils.cs (hunk 12 above)
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ConverterUtils.cs:101]: nevermind, I see where this is used in the product code path. I want to see if we can find a solution to this that doesn't involve regexes if we can avoid it.

Comment (rynowak) [reply src/Dapr.Actors/Runtime/ConverterUtils.cs:101]: @layomia - do you know if S.T.J or other parts of the runtime provide anything here that will help? I don't want to introduce new parsers or regexes that I don't trust 😆

Comment (layomia) [reply src/Dapr.Actors/Runtime/ConverterUtils.cs:101]: The ISO 8601 duration format doesn't actually map directly to our .NET `TimeSpan` type. `TimeSpan` is an absolute representation in that it counts number of ticks, while ISO duration is relative - allowing representations like "day", and "month". That's likely why it isn't included as part of the [standard formats for `TimeSpan`](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-timespan-format-strings).

STJ recently (and finally! 😄)  added support for `TimeSpan` (https://github.com/dotnet/runtime/pull/54186), but that supports only the "standard .NET" (`c`) format. So AFAIK, there are no built-in types to support this lossy conversion from ISO 8601 duration to .NET `TimeSpan`. We'd need a new type in the BCL to support the ISO format. This leaves us with custom parsers like with this PR.

File: src/Dapr.Actors/Resources/SR.resx (hunk 11 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Resources/SR.resx:178]: I will revert the changes to this file

File: src/Dapr.Actors/Dapr.Actors.csproj (hunk 10 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Dapr.Actors.csproj:3]: Most of the changes in file are formatting changes because I opened it in vscode. 
I have added InternalsVisibleToAttribute to ease mocking because some types are internal and

File: src/Dapr.Actors/Runtime/IActorReminder.cs (hunk 13 above)
Comment (jigargandhi) [reply src/Dapr.Actors/Runtime/IActorReminder.cs:49]: Its only read internally,

**Comment by yash-nisar** [comment]
Can be closed in favour of https://github.com/dapr/dotnet-sdk/pull/974. Thanks @jigargandhi for your work on this. :)

**Comment by halspang** [comment]
Closing as #974 handled this.
